scalus-core/shared/src/main/scala/scalus/cardano/ledger/value/multiasset/CanonicalSortedMap.scala

```diff
@@ -0,0 +1,235 @@
+  * the map itself can be an additive monoid, because [[SortedMap.empty]] can be the unique additive
+  * identity.
+  */
+type CanonicalSortedMap[K, V] = CanonicalSortedMap.CanonicalSortedMap[K, V]
+
+private object CanonicalSortedMap {
+    opaque type CanonicalSortedMap[K, V] = SortedMap[K, V]
+
+    /** Canonicalize a [[SortedMap]] by removing all keys corresponding to zero value. */
+    def apply[K, V](sortedMap: SortedMap[K, V])(using
+        vMonoid: AdditiveMonoid[V]
+    ): CanonicalSortedMap[K, V] =
+        sortedMap.filterNot(_._2 == vMonoid.zero)
+
+    def empty[K, V](using Ordering[K]): CanonicalSortedMap[K, V] = SortedMap.empty
+
+    extension [K, V](self: CanonicalSortedMap[K, V])
+        def underlying: SortedMap[K, V] = self
+
+        /** Apply a function to every value in the canonical sorted map. */
+        def mapValues[VNew](f: V => VNew)(using
+            kOrdering: Ordering[K],
+            vNewMonoid: AdditiveMonoid[VNew]
+        ): CanonicalSortedMap[K, VNew] = {
+            val newSortedMap: SortedMap[K, VNew] = self.view.mapValues(v => f(v)).to(SortedMap)
+            apply(newSortedMap)(using vMonoid = vNewMonoid)
+        }
+
+        /** Apply a function to every value in the canonical sorted map, exposing the value's key to
+          * the function as an additional argument.
+          */
+        def mapValuesIndexed[VNew](f: (K, V) => VNew)(using
+            kOrdering: Ordering[K],
+            vNewMonoid: AdditiveMonoid[VNew]
+        ): CanonicalSortedMap[K, VNew] = {
+            def fIndexed(kv: (K, V)): (K, VNew) = (kv._1, f(kv._1, kv._2))
+            val newSortedMap: SortedMap[K, VNew] = self.view.map(fIndexed).to(SortedMap)
+            apply(newSortedMap)(using vMonoid = vNewMonoid)
+        }
+
+    //    def combineWithSimple[K, V](
+    //        opBoth: (V, V) => V
+    //    )(self: CanonicalSortedMap[K, V], other: CanonicalSortedMap[K, V])(using
+    //        ev: AdditiveMonoid[V],
+    //        ord: Ordering[K]
+    //    ): CanonicalSortedMap[K, V] = {
+    //        (self.keySet ++ other.keySet).view
+    //            .map { key =>
+    //                val combinedValue =
+    //                    opBoth(self.getOrElse(key, ev.zero), other.getOrElse(key, ev.zero))
+    //                key -> combinedValue
+    //            }
+    //            .to(SortedMap)
+    //    }
+
+    def combineWith[K, VCommon, VResult, VSelf, VOther](
+        opBoth: (VCommon, VCommon) => VResult,
+        opSelf: VCommon => VResult = identity[VResult],
+        opOther: VCommon => VResult = identity[VResult],
+        preMapSelf: VSelf => VCommon = identity[VCommon],
+        preMapOther: VOther => VCommon = identity[VCommon]
+    )(
+        self: CanonicalSortedMap[K, VSelf],
+        other: CanonicalSortedMap[K, VOther]
+    )(using
+        kOrdering: Ordering[K],
+        vResultMonoid: AdditiveMonoid[VResult]
+    ): CanonicalSortedMap[K, VResult] = {
```

Comment:
I think the type signature would make more sense as

```suggestion
    def combineWith[K, VResult, VLHS, VRHS](
        opBoth: (Option[VLHS], Option[VRHS]) => Option[VResult],
    )(
        self: CanonicalSortedMap[K, VSelf],
        other: CanonicalSortedMap[K, VOther]
    )(using
        kOrdering: Ordering[K],
        vResultMonoid: AdditiveMonoid[VResult]
    ): CanonicalSortedMap[K, VResult] = {
```

which mirrors `updateWith` from map and seems much simpler.